pygmt/clib/session.py

```diff
@@ -1087,6 +1148,16 @@ def virtualfile_from_vectors(self, *vectors):
         arrays = vectors_to_arrays(vectors)
 
         columns = len(arrays)
+        # Find arrays that are of string dtype from column 3 onwards
+        # Assumes that first 2 columns contains coordinates like longitude
+        # latitude, or datetime string types.
+        str_cols = [
+            col + 2
+            for col, array in enumerate(arrays[2:])
+            if np.issubdtype(array.dtype, np.str_)
+        ]
+        columns -= len(str_cols)
```

Comment:
I've hardcoded things here so that the first two columns will never be passed into `put_strings`, since we can have `datetime` strings in the first 2 columns (x and y) since #464. Problem is that this hardcoding prevents having a `datetime` column on the 3rd column, which might affect `plot3d` (at #471) if someone decides to plot a time axis on the z-axis. I'm not sure how often anyone would plot time on the z-axis, but maybe we should find a way around this hardcoding.

Reply:
I don't think you recent changes make sense. If I understand your codes correctly, if someone pass arrays like the order below:
```
x = [0.5],
y = [0.5],
just = ["BL"],
angle = [90],
text = ["this is a label"]
```
you changes pass *x*, *y*, *angle* as the 0, 1 and 3 columns by calling *put_vectors*, and pass "BL this is a label" by calling `put_strings`. This won't work because column 2 is missing (I don't test it, because the **text** module actually don't allow pass *angle* after *justification*, but I think maybe there are other cases we may want to pass arrays like this).

Reply:
Yes I did think of this. I was thinking of modifying `lib.virtualfile_from_vectors` to explicitly have a `num_vectors` and `str_vectors` arguments, so we know which arrays are numerical and which are strings. This would affect a lot of code in the library though.

Reply:
Specifically, the `*extra_arrays` in `plot` wouldn't work if we change to something like `def virtualfile_from_vectors(num_arrays, str_arrays)`:

https://github.com/GenericMappingTools/pygmt/blob/62580f1f8de641ee14a21ae762dc207b3ae29595/pygmt/base_plotting.py#L553-L555

Reply:
When looping over arrays, we need to check the **first** np.str_ string. In the example above, it's *just*. Then we can pass all vectors before *just* as vectors, and combine the remaining arrays as one np.str_ array (although *angle* is numeric), and pass it to `put_strings`.

Reply:
> When looping over arrays, we need to check the **first** np.str_ string.

Yes, but we would still need to skip over the first 2 columns (or 3 columns in the case of `plot3d`) because datetimes can be string types. Our check for datetime isn't particularly nice since it tries to convert the whole array:

https://github.com/GenericMappingTools/pygmt/blob/62580f1f8de641ee14a21ae762dc207b3ae29595/pygmt/clib/session.py#L721-L729

> In the example above, it's _just_. Then we can pass all vectors before _just_ as vectors, and combine the remaining arrays as one np.str_ array (although _angle_ is numeric), and pass it to `put_strings`.

Ok, I can handle the 'combine remaining arrays' part, but is it okay to pass in `angle` to `put_strings` too?

Reply:
> Yes, but we would still need to skip over the first 2 columns (or 3 columns in the case of `plot3d`) because datetimes can be string types. Our check for datetime isn't particularly nice since it tries to convert the whole array:

Skipping the first 2 or 3 columns should work for most cases. If I understand it correctly, we can even pass all arrays as a single string, e.g., a single array `string = ["0.5 0.5 BL 90 this is a lable"]` should also work. Again, I haven't test it.

Reply:
Ok, I'll refactor the code to handle the 'combine remaining arrays', but will continue to skip the first 2 columns for now, until someone files a bug report for `plot` or `plot3d`.